### PR TITLE
Normative: add Iterator Helpers

### PR DESCRIPTION
**Pull request by michaelficarra**
https://github.com/tc39/proposal-iterator-helpers

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -6956,6 +6996,28 @@ <h1>
       </emu-alg>
     </emu-clause>
 
+    <emu-clause id="sec-getiteratorflattenable" type="abstract operation">
+      <h1>
+        GetIteratorFlattenable (
```
Comment (michaelficarra) [inline spec.html:7001]: I would love to come up with a better name for this AO.

**Comment by jmdyck** [comment]
Rather than add a bunch of suggestions to this PR, I've created PR #3398 against its branch.

**Review by jmdyck — DISMISSED** [review]

File: spec.html [hunk 2]
```diff
@@ -46436,27 +46520,465 @@ <h1>The <i>IteratorResult</i> Interface</h1>
...
+          1. Let _integerLimit_ be ! ToIntegerOrInfinity(_numLimit_).
+          1. If _integerLimit_ &lt; 0, throw a *RangeError* exception.
+          1. Let _iterated_ be ? GetIteratorDirect(_O_).
+          1. Let _closure_ be a new Abstract Closure with no parameters that captures _iterated_ and _integerLimit_ and performs the following steps when called:
+            1. Let _remaining_ be _integerLimit_.
+            1. Repeat, while _remaining_ > 0,
+              1. If _remaining_ is not +∞, then
+                1. Set _remaining_ to _remaining_ - 1.
+              1. Let _next_ be ? IteratorStep(_iterated_).
+              1. If _next_ is *false*, return *undefined*.
```
[the remark this answers is not in the document]
Comment (bakkot) [reply spec.html:46708]: Should compare against `~done~`, rather. This method advances the underlying iterator without reading its `.value`.

(`IteratorStep` returned `false` prior to https://github.com/tc39/ecma262/pull/3311 and I didn't update the proposal.)

[the remark this answers is not in the document]
Comment (michaelficarra) [reply spec.html:46708]:
```suggestion
              1. Let _next_ be ? IteratorStep(_iterated_).
              1. If _next_ is ~done~, return *undefined*.
```

**Comment by michaelficarra** [comment]
FYI we're having a relevant editorial discussion here: https://github.com/tc39/proposal-joint-iteration/pull/30

File: spec.html (hunk 1 above)
Comment (michaelficarra) [reply spec.html:7001]: This comment was about `GetIteratorFlattenable` (it's no longer displaying under the appropriate line in the diff view).

File: spec.html [hunk 3]
```diff
@@ -6888,6 +6909,32 @@ <h1>
           1. Set _k_ to _k_ + 1.
       </emu-alg>
     </emu-clause>
+
+    <emu-clause id="sec-SetterThatIgnoresPrototypeProperties" type="abstract operation">
+      <h1>
+        SetterThatIgnoresPrototypeProperties (
```
Comment (ljharb) [inline spec.html:6915]: does this AO belong in this PR?

**Review by ljharb — COMMENTED** [review]
This PR seems to be missing changes to CreateIteratorFromClosure to add a 4th argument? (present in the proposal spec)

File: spec.html (hunk 3 above)
Comment (syg) [reply spec.html:6915]: Well, it's used in this PR, and we generally don't land AOs that have no users.

File: spec.html [hunk 4]
```diff
@@ -6888,6 +6909,32 @@ <h1>
...
+          _this_: an ECMAScript language value,
```
Comment (syg) [inline spec.html:6916]:
```suggestion
          _thisValue_: an ECMAScript language value,
```

Given that we say things like "the this value", it'd be best to not have an alias with the same name.

File: spec.html [hunk 5]
```diff
@@ -46491,27 +46575,465 @@ <h1>The <i>IteratorResult</i> Interface</h1>
...
+          <emu-alg>
+            1. Let _O_ be *this* value.
+            1. Perform ? RequireInternalSlot(_O_, [[UnderlyingIterator]]).
+            1. Assert: _O_ has a [[GeneratorState]] internal slot.
+            1. If _O_.[[GeneratorState]] is ~suspended-start~, then
+              1. Set _O_.[[GeneratorState]] to ~completed~.
+              1. NOTE: Once a generator enters the completed state it never leaves it and its associated execution context is never resumed. Any execution state associated with _O_ can be discarded at this point.
+              1. Perform ? IteratorClose(_O_.[[UnderlyingIterator]], ReturnCompletion(*undefined*)).
+              1. Return CreateIteratorResultObject(*undefined*, *true*).
+            1. Let _C_ be Completion Record { [[Type]]: ~return~, [[Value]]: *undefined*, [[Target]]: ~empty~ }.
```
Comment (syg) [inline spec.html:46610]: ReturnCompletion?

**Review by syg — APPROVED** [review]

File: spec.html (hunk 3 above)
Comment (ljharb) [reply spec.html:6915]: ah, sorry, i missed the usage and apparently incorrectly recalled that we'd decided on a different approach than this.

**Comment by michaelficarra** [comment]
@tc39/ecma262-editors Do we want to make the `[[UnderlyingIterator]]` internal slot plural (as is required by upcoming proposal joint iteration) as part of this PR, or should we do it in 2 phases?

**Review by bakkot — APPROVED** [review]
LGTM. I no longer think we should rename `[[UnderlyingIterators]]` now, since it would also involve changing its handling; we can do that in the followup.

File: spec.html [hunk 6]
```diff
@@ -46491,27 +46575,465 @@ <h1>The <i>IteratorResult</i> Interface</h1>
...
+        <emu-clause id="sec-%iteratorhelperprototype%.return">
+          <h1>%IteratorHelperPrototype%.return ( )</h1>
+          <emu-alg>
+            1. Let _O_ be *this* value.
+            1. Perform ? RequireInternalSlot(_O_, [[UnderlyingIterator]]).
+            1. Assert: _O_ has a [[GeneratorState]] internal slot.
+            1. If _O_.[[GeneratorState]] is ~suspended-start~, then
+              1. Set _O_.[[GeneratorState]] to ~completed~.
+              1. NOTE: Once a generator enters the completed state it never leaves it and its associated execution context is never resumed. Any execution state associated with _O_ can be discarded at this point.
+              1. Perform ? IteratorClose(_O_.[[UnderlyingIterator]], ReturnCompletion(*undefined*)).
```
Comment (jmdyck) [inline spec.html:46608]: It seems like this Return Completion could be returned by `IteratorClose`, and thus by this algorithm, and thus by the `[[Call]]` for this built-in. But the invariants for essential internal methods don't allow `[[Call]]` to return a Return Completion.

File: spec.html [hunk 7]
```diff
@@ -3658,12 +3658,23 @@ <h1>Well-Known Intrinsic Objects</h1>
             </tr>
             <tr>
               <td>
-                %IteratorPrototype%
```
Comment (ljharb) [inline spec.html:3661]: i missed this removal; the intention was to keep pre-existing intrinsics in this table. i'll make a PR to re-add `%IteratorPrototype%` in the future.

**Pull request merged**